### ISSUE TITLE
In-place ops in selective checkpointing region are not preserved in AOTAutograd joint graph

### ISSUE THREAD

**Issue by yf225** [pytorch/pytorch#113737]
### 🐛 Describe the bug

As titled. To repro:
1. Remove the `@unittest.skip` block on top of
`test/dynamo/test_activation_checkpointing.py::ActivationCheckpointingViaTagsTests::test_compile_selective_checkpoint_inplace_op`
(code pointer: https://github.com/pytorch/pytorch/blob/9724d0fd87af6107a22dc06d7a7fda5a26fa0615/test/dynamo/test_activation_checkpointing.py#L578)
2. Run:
`TORCH_COMPILE_DEBUG=1 pytest -vs test/dynamo/test_activation_checkpointing.py::ActivationCheckpointingViaTagsTests::test_compile_selective_checkpoint_inplace_op`
3. We get this log: https://gist.github.com/yf225/0e8ebcbaa4aa05db1a80e18e51a7d4d1. Notice that `__compiled_fn_0` has the in-place ops (as expected), but `Joint graph 0` doesn't have the in-place ops (which seems not correct).

### Versions

PyTorch nightly

**Comment by yf225** [comment]
Working with Brian on this.

**Comment by bdhirsh** [comment]
So far I'm just working on making the repro a bit smaller. This still fails:
```
        def gn(x, y):
            return torch.selu_(torch.matmul(x, y))

        def fn(x, y):
            return torch.utils.checkpoint.checkpoint(
                gn,
                x,
                y,
                use_reentrant=False,
                context_fn=selective_checkpointing_context_fn,
            )
```

**Comment by bdhirsh** [comment]
I spent some time creating a repro, here's a smaller reproduction:
```
import torch
from torch._custom_op.functional import register_functional_op
import torch.utils.checkpoint
from torch.utils.checkpoint import checkpoint, context_fn_gen

def custom_policy(mode, func, *args, **kwargs):
    return func in [torch.ops.aten.mm.default]

def selective_checkpointing_context_fn():
    return context_fn_gen(custom_policy)

def gn(x, y):
    return torch.selu_(torch.matmul(x, y))

def fn(x, y):
    return torch.utils.checkpoint.checkpoint(
        gn,
        x,
        y,
        use_reentrant=False,
        context_fn=selective_checkpointing_context_fn,
    )

x = torch.arange(16, dtype=torch.float32, requires_grad=True).reshape(4, 4).detach().requires_grad_(True)
y = torch.arange(16, dtype=torch.float32, requires_grad=True).reshape(4, 4).detach().requires_grad_(True)

out = fn(x, y)
print(out)
# Invoking the backward will mutate out!
out.sum().backward()
print(out)

torch._dynamo.config._experimental_support_context_fn_in_torch_utils_checkpoint = True
out = torch.compile(fn)(x, y)
print(out)
# But torch.compile drops this mutation in the backward
out.sum().backward()
print(out)
```

This is pretty interesting: if you print the forward out before/after calling `out.sum().backward()`, you'll see that in eager mode, we end up **mutating** the forward output when we run the backward.

In the above example, this is because the forward output is both an output to the forward, and one of the tensors that gets saved for backward (because it's the result of an mm, which is in the SAC norecompute list)

`selu_()` is not in the norecompute list, so we will re-run `selu_()` in both the forward and the backward. But since the input to the `selu_()` was a saved-for-backward tensor, we will end up mutating it during the recompute part of SAC.

This feels relatively similar to https://github.com/pytorch-labs/float8_experimental/issues/106#issuecomment-1791737080: There's a mutation that happens during the backward pass, that is observable to the user (it mutates one of the inputs/outputs to the forward). So a few questions are:

(1) Should we actually try to handle this in torch.compile? Or should we figure out how to detect it, and error out (but let compiled autograd handle it).

In order to detect it, we need to know that (a) we're running an inplace op inside the SAC region, and (b) the input to the inplace op is either a forward graph input or forward graph output. It's not immediately clear to me if we can detect this. Maybe we can settle for (pessimistically) error'ing if we see that we're doing an inplace op on an input that we saved_for_backward? (cc @soulitzer in case you have any thoughts)

(2) How rare of a case to we expect this to be? It feels kind of rare, but to be fair, it's probably possible for a user to construct code like this without realizing that they're hitting this case.

Marking triage_review to discuss priority

**Comment by yf225** [comment]
To make it easier, we can also just ban in-place ops in compile+SAC. In-place ops should be pretty rare in general, and I believe we can wait for compiled autograd to be ready to support in-place ops in compile+SAC.

One downside is that full activation checkpoint (AC) does support in-place ops. So this will prevent us from merging the compiled impl of AC and SAC into one.

cc. @Chillee @anijain2305

**Comment by ezyang** [comment]
My vote is still for compiled autograd lol

**Comment by bdhirsh** [comment]
I think I have more detailed example / writeup of the problem (it's not related to functionalization / compiled autograd). The tldr is:

(1) The torch.compile implementation of SAC works by tracing out a joint fw/bw graph that does *zero* recomputation, but instead tags the nodes in the forward that should be recomputed. It then relies on the partitioner to look at these tags when choosing what to recompute vs. save for backward

(2) This relies on the assumption that recomputing vs. saving for backward has no effect on the output

(3) However, in eager mode it is possible to get **different results** when you wrap regions of code with SAC, which breaks the above assumption.

Here's some code that demonstrates how you can get different gradients depending on how you use SAC in eager mode:
```
def f(x):
    y = torch.matmul(x, x)
    y2 = torch.elu_(y, ...)
    z = torch.sigmoid(z)
    return z
```

`sigmoid` (and also `elu`/`elu_`) saves its output for the backward pass. So the backward graph (without any selective activation checkpointing) looks something like:
```
# saved_tensors_for_backward:
# - z (for sigmoid_backward)
# - y2 (for elu_backward)
# - x (for the matmul_backward)
def backward_graph(grad_out, z, y2, x):
    # sigmoid_backward saves the output of sigmoid() from the forward
    sigmoid_grad = sigmoid_backward(grad_out, z)
    # elu_backward saves the output of elu_() from the forward
    elu_grad = elu_backward(sigmoid_grad, ..., y2)
    ... remainder of the backward
```

With SAC though, we can recompute everything in `f()` except for the matmuls. If we blindly recompute everything except the matmuls, we'll have:
```
# saved_tensors_for_backward:
# - y (output of the matmul is saved for backward, since it's needed to recompute elu and we don't recompute the matmul)
# - x (for the matmul_backward)
def backward_graph_recompute_selu_and_sigmoid(grad_out, y, x):
    # recompute of elu_()
    y2 = elu_(y)
    # recompute sigmoid()
    z = sigmoid(y2)
    # sigmoid backward, uses the recomputed sigmoid output in the backward
    sigmoid_grad = sigmoid_backward(grad_out, z)
    # elu_backward, uses the recomputed elu_ output in the backward
    elu_grad = elu_backward(sigmoid_grad, ..., y2)
    ... remainder of the backward
```

These two functions return **different** gradients though. In particular, the "y2" that we pass to `sigmoid_backward()` is the result of `matmul(x, x)`, and calling `elu_()` twice on it (once in the forward again in the backward). While in the first example, we only called `elu_()` on it once (in the forward), before "freezing" it when we saved it for backward.

So I guess my question is for @soulitzer- is it intended behavior that you can use SAC in eager mode in a way that will change the computed gradients (or even the forward output)? And if not, should we keep the behavior around or would it be possible to detect/ban it. I'm not sure how easy this is, since it's not as simple as "ban all mutations in an SAC region". This only happens when we have a tensor used in the backward that we can choose to recompute (from some other saved activation), and the act of recomputing it mutates that other saved activation, so you get a different result every time your recompute it.

**Comment by soulitzer** [comment]
Thanks for the thorough investigation!

SAC is not actually supported in eager mode. It is a feature specific to compile today as we don't have sufficient infra in order to ensure correctness in a nice way in eager. We do offer the context_fn API, so users are free to whatever context_fn they'd like, but that is at their own risk. 

What may be confusing today is that we do provide such a context_fn in core. We should make it clear that it is intended for use with torch.compile only.
